src/gpu/generic/sycl/ref_reduction.cpp

```diff
@@ -0,0 +1,479 @@
+        auto dims_begin = dims.begin();
+        auto dims_end = dims.end();
+        auto axis = axes[red_iter];
+        conf.is_first_iter = (red_iter == 0);
+        conf.is_last_iter = (red_iter == num_reductions_ - 1);
+        conf.batch_size = std::accumulate(
+                dims_begin, dims_begin + axis, 1, std::multiplies<int>());
+        conf.reduce_size = dims[axis];
+        if (axis < static_cast<int>(dims.size() - 1)) {
+            conf.stride_size = std::accumulate(
+                    dims_begin + axis + 1, dims_end, 1, std::multiplies<int>());
+        } else {
+            conf.stride_size = 1;
+        }
+
+        needs_atomic_reduction_ = conf.batch_groups != -1;
+        const auto batch_groups
+                = conf.batch_groups == -1 ? conf.batch_size : conf.batch_groups;
+        const auto max_wg_size = max_wg_size_;
+        const auto max_sg_size = max_sg_size_;
+        int tile_col = get_max_col_tile_dim(
+                conf.stride_size, max_wg_size, max_sg_size);
+        tile_col = std::min(tile_col, sycl_reduction_conf_t::local_col_wg);
+        int tile_row = get_max_row_tile_dim(
+                conf.reduce_size, tile_col, max_wg_size, max_sg_size);
+        tile_row = std::min(tile_row, sycl_reduction_conf_t::local_row_wg);
+        conf.tile_col = tile_col;
+        conf.tile_row = tile_row;
+        local_ranges_.emplace_back(range_t {1, tile_row, tile_col});
+
+        auto global_col
+                = round_up_to_nearest_multiple(conf.stride_size, tile_col);
+        auto global_row
+                = round_up_to_nearest_multiple(conf.reduce_size, tile_row);
+        global_ranges_.emplace_back(
+                range_t {conf.batch_size, global_row, global_col});
+        needs_atomic_reduction_ = needs_atomic_reduction_
+                || (global_row > std::min(
+                            tile_row, conf.num_sg_reductions * max_sg_size));
+
+        if (needs_atomic_reduction_ && conf.alg == dnnl_reduction_mul) {
+            return status::unimplemented;
+        }
+        if (needs_atomic_reduction_ && attr()->post_ops_.find(dnnl_sum) != -1) {
+            return status::unimplemented;
+        }
+
+        const size_t dt_size = data_type_size(dnnl_f32);
+        local_mem_sizes_.push_back(
+                ((tile_row + conf.bank_offset) * (tile_col + conf.bank_offset))
+                * dt_size);
+
+        needs_reorder_
+                = needs_atomic_reduction_ && dst_wrap.data_type() != dnnl_f32;
+        if (needs_reorder_) { conf.dst_dt = dnnl_f32; }
+
+        if (multi_reduction_) {
+            if (red_iter != 0) { conf.src_dt = dnnl_f32; }
+            if (red_iter != num_reductions_ - 1) { conf.dst_dt = dnnl_f32; }
+        }
+
+        conf.batch_groups = batch_groups;
+        confs_.push_back(conf);
+        dims[axes[red_iter]] = 1;
+    }
+
+    if (needs_reorder_) { CHECK(init_reorder(engine)); }
+    if (needs_atomic_reduction_) { CHECK(init_out_scratchpad()); }
```

Comment:
Can we check that the SYCL device supports atomic when `needs_atomic_reduction_` is `true`? I think we just need to check the query `info::device::atomic_memory_scope_capabilities` contains `memory_scope::work_group`.

Reply:
Good point, done.